### PR TITLE
 ✨ feat(tests): EIP-7928 tests for coinbase, withdrawal, storage, and account access

### PR DESCRIPTION
**Pull request by raxhvl**
Adds new EIP-7928 test cases for Coinbase, withdrawal, edge cases for storage and account access.

### Coinbase
  1. `test_bal_empty_block_no_coinbase` 
  2. `test_bal_coinbase_zero_tip`

 ### EIP-4895 Withdrawal
 
1. `test_bal_withdrawal_empty_block`
2. `test_bal_withdrawal_and_transaction`
3. `test_bal_withdrawal_to_nonexistent_account`
4. `test_bal_withdrawal_no_evm_execution`
5. `test_bal_withdrawal_and_state_access_same_account`
6. `test_bal_withdrawal_and_value_transfer_same_address`
7. `test_bal_multiple_withdrawals_same_address`
8. `test_bal_withdrawal_and_selfdestruct`
9. `test_bal_withdrawal_and_new_contract`
10. `test_bal_zero_withdrawal`
11. `test_bal_withdrawal_to_precompiles`
12. `test_bal_withdrawal_largest_amount`
13. `test_bal_withdrawal_to_coinbase`
14. `test_bal_withdrawal_to_coinbase_empty_block`

## Precompiles
1. `test_bal_precompile_funded`
2. `test_bal_precompile_call`

## Non-existent accounts
1. `test_bal_nonexistent_value_transfer`
2. `test_bal_nonexistent_account_access`

## Storage edge cases
1. `test_bal_storage_write_read_same_frame`
2. `test_bal_storage_write_read_cross_frame`

## Blockers

- [x] https://github.com/fselmo/execution-specs/issues/24
- [x] https://github.com/fselmo/execution-specs/issues/26
- [x] https://github.com/fselmo/execution-specs/issues/27
- [x] https://github.com/fselmo/execution-specs/issues/30

### CODE REVIEW

**Comment by raxhvl** [comment]
Coinbase test blocked by https://github.com/fselmo/execution-specs/issues/24

File: tests/amsterdam/eip7928_block_level_access_lists/test_cases.md [hunk 1]
```diff
@@ -58,5 +58,19 @@
...
+| `test_bal_withdrawal_and_new_contract` | Ensure BAL captures withdrawal to newly created contract | Block with 1 transaction: Alice deploys contract `Oracle` with 5 gwei initial balance. 1 withdrawal of 10 gwei to `Oracle` | BAL **MUST** include `Oracle` with `code_changes` and `balance_changes` showing 5 gwei at `block_access_index = 1`. `Oracle` **MUST** also have `balance_changes` at `block_access_index = 2` showing balance after receiving 10 gwei withdrawal. `Oracle`'s final post-state balance **MUST** be 15 gwei (cumulative). | ✅ Completed |
+| `test_bal_zero_withdrawal` | Ensure BAL handles zero-amount withdrawal correctly | Block with 0 transactions and 1 zero-amount withdrawal (0 gwei) to Charlie. Two variations: Charlie has existing balance (5 gwei) or Charlie is non-existent. | BAL **MUST** include Charlie at `block_access_index = 1` with empty changes. Balance remains unchanged. | ✅ Completed |
```
Comment (fselmo) [inline tests/amsterdam/eip7928_block_level_access_lists/test_cases.md:72]: I need clarity here. A `0` value transfer to a non-existent account is still marked as a read right? Just sanity checking @nerolation @raxhvl

Comment (fselmo) [reply tests/amsterdam/eip7928_block_level_access_lists/test_cases.md:72]: Sorry, it was a different issue I was seeing where the test was wrongly expecting a `0` in `post` instead of a nonexistent account. Fixed [here](https://github.com/fselmo/execution-specs/pull/23/commits/203763d1e9b94b5b574c9cdfbc1ded5d79583f25).

**Review by fselmo — APPROVED** [review]
@raxhvl I added some fixes to this but this has everything passing after #29 gets in. I will wait for that and then we can merge this. Thanks for all the work here! 👍🏼

File: tests/amsterdam/eip7928_block_level_access_lists/test_cases.md (hunk 1 above)
Comment (raxhvl) [reply tests/amsterdam/eip7928_block_level_access_lists/test_cases.md:72]: > I need clarity here. A `0` value transfer to a non-existent account is still marked as a read right?

Read with empty changes.

**Comment by fselmo** [comment]
@raxhvl can you fix the `static` lint issues here?

**Comment by raxhvl** [comment]
Hey @fselmo fixed the static issue, I have few more new tests since your review:

:sparkle: 

## Withdrawal
11. `test_bal_withdrawal_to_precompiles`
12. `test_bal_withdrawal_largest_amount`
13. `test_bal_withdrawal_to_coinbase`
14. `test_bal_withdrawal_to_coinbase_empty_block`

## Precompiles
1. `test_bal_precompile_funded`
2. `test_bal_precompile_call`

## Non-existent accounts
1. `test_bal_nonexistent_value_transfer`
2. `test_bal_nonexistent_account_access`

## Storage edge cases
1. `test_bal_storage_write_read_same_frame`
2. `test_bal_storage_write_read_cross_frame`


I'm done adding tests now - Once all issues are closed feel free to merge the tests if everything passes. Could you also carefully review the coinbase once before merging? Thanks!

**Comment by fselmo** [comment]
@raxhvl this looks good to me now. I fixed the issues with the precompile tests [here](https://github.com/fselmo/execution-specs/pull/23/commits/7c0a150ec3b33dd2be6138039c1de6299d8236a3) (the tests were wrong, not the specs) and I split the test that was using `pytest.skip` because I don't think we should have this built into tests. Please make sure this looks good to you [here](https://github.com/fselmo/execution-specs/pull/23/commits/4906edaec6cd69015a8264eb75e7e249f9c7ba16) and if so I'm good to merge this so we can move this feature branch over to `ethereum/execution-specs` as branch `eips/amsterdam/eip-7928` so we can start to release from there. The reason for this is we need the feature branch to be part of `ethereum/execution-specs` rather than my fork (`fselmo/execution-specs`).

**Comment by raxhvl** [comment]
Hey @fselmo - The changes look good to me, all tests are now passing, you can merge this :+1:

**Comment by fselmo** [comment]
@raxhvl I've created the branch to execution-specs [here](https://github.com/ethereum/execution-specs/pull/1719) with these changes in there! Going to close this as we're all caught up there now.
